jira/resilientsession.py

```diff
@@ -100,20 +104,137 @@ def __init__(self, timeout=None):
         # Indicate our preference for JSON to avoid https://bitbucket.org/bspeakmon/jira-python/issue/46 and https://jira.atlassian.com/browse/JRA-38551
         self.headers.update({"Accept": "application/json,*.*;q=0.9"})
 
+    def _jira_prepare(self, **_kwargs) -> dict:
+        """Do any pre-processing of our own and return the updated kwargs."""
+        kwargs = _kwargs.copy()
+
+        request_headers = self.headers.copy()
+        request_headers.update(_kwargs.get("headers", {}))
+        kwargs["headers"] = request_headers
+
+        data = _kwargs.get("data", {})
+        if isinstance(data, dict):
+            # mypy ensures we don't do this,
+            # but for people subclassing we should preserve old behaviour
+            kwargs["data"] = json.dumps(data)
+
+        return kwargs
+
+    def request(  # type: ignore[override] # An intentionally different override
+        self,
+        method: str,
+        url: Union[str, bytes],
+        _prepare_retry_method: PrepareRequestForRetry = PassthroughRetryPrepare(),
+        **_kwargs,
+    ) -> Response:
+        """This is an intentional override of `Session.request()` to inject some error handling and retry logic.
+
+        Raises:
+            Exception: Various exceptions as defined in py:method:`raise_on_error`.
+
+        Returns:
+            Response: The response.
+        """
+
+        retry_number = 0
+        exception: Optional[Exception] = None
+        response: Optional[Response] = None
+        response_or_exception: Optional[Union[ConnectionError, Response]]
+
+        processed_kwargs = self._jira_prepare(**_kwargs)
+
+        def is_retrying() -> bool:
+            """Helper method to say if we should still be retrying."""
+            return retry_number <= self.max_retries
+
+        while is_retrying():
+            response = None
+            exception = None
+
+            try:
+                response = super().request(method, url, **processed_kwargs)
+                if response.ok:
+                    self.__handle_known_ok_response_errors(response)
+                    return response
+            # Can catch further exceptions as required below
+            except ConnectionError as e:
+                exception = e
+
+            # Decide if we should keep retrying
+            response_or_exception = response if response is not None else exception
+            retry_number += 1
+            if is_retrying() and self.__recoverable(
+                response_or_exception, url, method.upper(), retry_number
+            ):
+                _prepare_retry_method.prepare(processed_kwargs)
+            else:
+                retry_number = self.max_retries + 1  # exit the while loop, as above max
+
+        if exception is not None:
+            # We got an exception we could not recover from
+            raise exception
+        elif raise_on_error(response, **processed_kwargs):
```

Comment:
I'm not too sure about the rationale of having the function `raise_on_error` returning a boolean and using it in an if condition: couldn't it follow the same behaviour of the `raise_for_status` function from requests.Response? Raise an exception if some conditions are met, otherwise no operation. Then you could do
```
raise_on_error(response)
return response
```
outside of the if statement, and you wouldn't need the Runtime exception (not 100% sure about this last statement)

Reply:
the reasoning is based around the static type checking, the [`TypeGuard` type](https://peps.python.org/pep-0647), it is basically instructing the type checker that we have checked that the `Response` is actually a `Response` and not `None`.

So we could definitely do what you described, but mypy will complain that `Optional[Response]` not a `Response` is being output from the function.
We could do a `cast()` to just tell mypy we know that it will be a `Response`, but I think this can let bugs slip through. 

Finally, `raise_on_error()` is automatically called for each request (unlike `raise_for_status()`) so it is more of an implementation detail than a user function, so I'm not too worried about how a user will use it.

Reply:
Ok, thanks for the explanation. It doesn't bother much, leave it as is and consider this resolved.